client/client.go

```diff
@@ -0,0 +1,322 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sort"
+	"strings"
+)
+
+// DefaultLogger is a default func to print logs
+func DefaultLogger(format string, args ...interface{}) {
+	log.Printf("[DEBUG] "+format, args...)
+}
+
+// RoundTripper satisfies the http.RoundTripper interface and is used to
+// customize the default http client RoundTripper
+type RoundTripper struct {
+	// Default http.RoundTripper
+	Rt http.RoundTripper
+	// Additional request headers to be set (not appended) in all client
+	// requests
+	headers *http.Header
+	// A pointer to a map of headers to be masked in logger
+	maskHeaders *map[string]struct{}
+	// A custom function to format and mask JSON requests and responses
+	FormatJSON func([]byte) (string, error)
+	// How many times HTTP connection should be retried until giving up
+	MaxRetries int
+	// If Logger is not nil, then RoundTrip method will debug the JSON
+	// requests and responses
+	Logger func(string, ...interface{})
+}
+
+// List of headers that contain sensitive data.
+var defaultSensitiveHeaders = map[string]struct{}{
```

Comment:
I've never seen `map[string]struct` used this way. Is there a benefit over doing this than `map[string]interface` and having the values set to `nil`?

Reply:
I used map because it's implemented over hash table and has O(1) in average so it's simpler and faster to check for keys in it instead of looping over a slice which gives us O(n) time.
https://dave.cheney.net/2018/05/29/how-the-go-runtime-implements-maps-efficiently-without-generics

Using `struct{}` instead of `interface{}` is common when you implement some signalling over channels when you don't need a value.

Empty struct has zero memory consumption vs empty interface. Consider this example:
```
package main

import (
    "fmt"
    "unsafe"
)

func main() {
    var s struct{}
    fmt.Println(unsafe.Sizeof(s))
    var i interface{}
    fmt.Println(unsafe.Sizeof(i))
}
// Prints:
// 0
// 8
```

https://dave.cheney.net/2014/03/25/the-empty-struct

Reply:
Right, I understand the use of a map. It's the `map[struct]struct` that I've never seen before. I'm fine with keeping it.